pymc/gp/hsgp.py

```diff
@@ -0,0 +1,376 @@
+        share the same basis.  The return values are the Laplace eigenfunctions `phi`, and the
+        square root of the power spectral density.
+
+        Correct results when using `prior_linearized` in tandem with `pm.set_data` and
+        `pm.MutableData` require two conditions.  First, one must specify `L` instead of `c` when
+        the GP is constructed.  If not, a RuntimeError is raised.  Second, the `Xs` needs to be
+        zero centered, so it's mean must be subtracted.  An example is given below.
+
+        Parameters
+        ----------
+        Xs: array-like
+            Function input values.  Assumes they have been mean subtracted or centered at zero.
+
+        Examples
+        --------
+        .. code:: python
+
+            # A one dimensional column vector of inputs.
+            X = np.linspace(0, 10, 100)[:, None]
+
+            with pm.Model() as model:
+                eta = pm.Exponential("eta", lam=1.0)
+                ell = pm.InverseGamma("ell", mu=5.0, sigma=5.0)
+                cov_func = eta**2 * pm.gp.cov.ExpQuad(1, ls=ell)
+
+                # m = [200] means 200 basis vectors for the first dimenison
+                # L = [10] means the approximation is valid from Xs = [-10, 10]
+                gp = pm.gp.HSGP(m=[200], L=[10], cov_func=cov_func)
+
+                # Order is important.  First calculate the mean, then make X a shared variable,
+                # then subtract the mean.  When X is mutated later, the correct mean will be
+                # subtracted.
+                X_mu = np.mean(X, axis=0)
+                X = pm.MutableData("X", X)
+                Xs = X - X_mu
+
+                # Pass the zero-subtracted Xs in to the GP
+                phi, sqrt_psd = gp.prior_linearized(Xs=Xs)
+
+                # Specify standard normal prior in the coefficients.  The number of which
+                # is given by the number of basis vectors, which is also saved in the GP object
+                # as m_star.
+                beta = pm.Normal("beta", size=gp.m_star)
+
+                # The (non-centered) GP approximation is given by
+                f = pm.Deterministic("f", phi @ (beta * sqrt_psd))
+
+                ...
+
+
+            # Then it works just like a linear regression to predict on new data.
+            # First mutate the data X,
+            x_new = np.linspace(-10, 10, 100)
+            with model:
+                model.set_data("X", x_new[:, None])
+
+            # and then make predictions for the GP using posterior predictive sampling.
+            with model:
+                ppc = pm.sample_posterior_predictive(idata, var_names=["f"])
+        """
+
+        # Index Xs using input_dim and active_dims of covariance function
+        Xs, _ = self.cov_func._slice(Xs)
+
+        # If not provided, use Xs and c to set L
+        S, self.L, self.c = set_boundaries(Xs, self.L, self.c, tl=pt)
```

Comment:
The `set_boundaries` returns a `ndarray` or `TensorVariable` for `L`, but the `__init__` signature defined it as a `Sequence[float]`. This gives errors because `ndarray/TensorVariable` don't qualify as `Sequence[float]`.

I would recommend to make `__init__` take `S: Optional[Union[Sequence[float], np.ndarray, TensorVariable]]` but inside `__init__` (`if S is not None`) apply `as_tensor` or `np.asarray` such that `self.S: Optional[Union[TensorVariable, np.ndarray]]`.

You can make the code more readable by aliasing `TensorLike: TypeAlias = Union[TensorVariable, np.ndarray]` somewhere at the top of the file.

Reply:
marking resolved since the mypy stuffs cleaned up now.   Switched to the name `TensorLike` too